auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt

```diff
@@ -0,0 +1,275 @@
+                    .log("Adding user {} to access group {}", session.userCn, action.getActiveDirectoryString())
+                groupService.addUserToGroup(
+                    callingUser.cn,
+                    callingUser.dn,
+                    action.getActiveDirectoryString(),
+                    call,
+                    null
+                )
+            } else if (action is RemoveAccessGroupAction || action is RemoveAccessGroupOrganisationAction) {
+                logger.atInfo()
+                    .log("Removing user {} from access group {}", session.userCn, action.getActiveDirectoryString())
+                groupService.removeUserFromGroup(
+                    callingUser.cn,
+                    callingUser.dn,
+                    action.getActiveDirectoryString(),
+                    call,
+                    null
+                )
+            }
+        }
+    }
+
+    private fun stripDatamartPrefixFromKeys(prefixedMap: Map<String, List<String>>): Map<String, List<String>> {
+        val newMap = mutableMapOf<String, List<String>>()
+        for (entry in prefixedMap.entries) {
+            newMap[entry.key.removePrefix(LDAPConfig.DATAMART_DELTA_PREFIX)] = entry.value
+        }
+        return newMap.toMap()
+    }
+
+    sealed class AccessGroupAction(
+        val accessGroupName: String,
+        val organisationCode: String?,
+    ) {
+        fun getActiveDirectoryString() = LDAPConfig.DATAMART_DELTA_PREFIX + accessGroupName +
+                if (organisationCode == null) "" else "-$organisationCode"
+    }
+
+    open class AddAccessGroupOrganisationAction(accessGroupName: String, organisationCode: String?) :
+        AccessGroupAction(accessGroupName, organisationCode) {
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            if (other !is AddAccessGroupOrganisationAction) return false
+
+            return this.accessGroupName == other.accessGroupName && this.organisationCode == other.organisationCode
+        }
+
+        override fun hashCode(): Int {
+            return javaClass.hashCode()
+        }
+    }
+
+    class AddAccessGroupAction(accessGroupName: String) :
+        AddAccessGroupOrganisationAction(accessGroupName, null)
+
+    open class RemoveAccessGroupOrganisationAction(accessGroupName: String, organisationCode: String?) :
+        AccessGroupAction(accessGroupName, organisationCode) {
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            if (other !is RemoveAccessGroupOrganisationAction) return false
+
+            return this.accessGroupName == other.accessGroupName && this.organisationCode == other.organisationCode
+        }
+
+        override fun hashCode(): Int {
+            return javaClass.hashCode()
+        }
+    }
+
+    class RemoveAccessGroupAction(accessGroupName: String) :
+        RemoveAccessGroupOrganisationAction(accessGroupName, null)
```

Comment:
Can pull the `equals` up to `AccessGroupAction` and compare Java clas, and should implement hashCode properly if we're going to do it.

```suggestion
    sealed class AccessGroupAction(
        val accessGroupName: String,
        val organisationCode: String?,
    ) {
        fun getActiveDirectoryString() = LDAPConfig.DATAMART_DELTA_PREFIX + accessGroupName +
                if (organisationCode == null) "" else "-$organisationCode"

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is AccessGroupAction) return false
            if (this.javaClass != other.javaClass) return false

            return this.accessGroupName == other.accessGroupName && this.organisationCode == other.organisationCode
        }

        override fun hashCode(): Int {
            var result = javaClass.hashCode()
            result = 31 * result + accessGroupName.hashCode()
            result = 31 * result + (organisationCode?.hashCode() ?: 0)
            return result
        }
    }

    class AddAccessGroupOrganisationAction(accessGroupName: String, organisationCode: String) :
        AccessGroupAction(accessGroupName, organisationCode)

    class AddAccessGroupAction(accessGroupName: String) :
        AccessGroupAction(accessGroupName, null)

    class RemoveAccessGroupOrganisationAction(accessGroupName: String, organisationCode: String) :
        AccessGroupAction(accessGroupName, organisationCode)

    class RemoveAccessGroupAction(accessGroupName: String) :
        AccessGroupAction(accessGroupName, null)
```